### ISSUE TITLE
In-depth profiling of pipeline steps

### ISSUE THREAD

**Issue by tskisner** [desihub/desispec#297]
Although we have done this kind of activity in the past, we need to revisit detailed profiling of PSF estimation, spectral extraction, and redshift fitting.  The starting point will be the current codebase, although I know people are actively working on improvements to those steps.

**Comment by tskisner** [comment]
As part of this profiling, also do a test extraction with HARP for comparison.

**Comment by tskisner** [comment]
This profiling should be done with more definitive productions and versions, but I have run the pine production through by using the master version of desispec and tweaking the redshift slurm script to give it a longer run time.  This gives us the following rough numbers:

Pine production:  10 spectrographs for 3 arcs, 3 flats, and 24 science exposures.

|   Step      | Nodes | Python Start |   Run Time |
| ------------ | ------- | -------------- | ------------- |
| bootstrap    |     3 |   0m : 11s   |   10m : 25s |
| psf          |   150 |   0m : 44s   |   14m : 44s |
| psfcombine   |     2 |   0m : 15s   |    0m : 39s |
| extract      |   675 |   5m : 11s   |   14m : 38s |
| calibrate    |    30 |   0m : 41s   |    5m : 22s |
| bricks       |     1 |   Unknown    |   52m : 26s (A) |
| redshift     |   168 |   1m : 31s   |  343m : 03s (B) |


(A) Previously the make_bricks.py script kept all brick files open when reading the cframe files and writing spectra.  For a large number of bricks, this exceeds available RAM.  The code was modified to open/close only the bricks needed for each cframe file.  This is slower but reduces the RAM needed.  We should come up with a new way of organizing this information, which would also make it easier to load balance the redshift fitting.

(B) The brick files contain 2-600 targets.  Since our MPI wrapped version of redmonster works on one brick at a time, it is a tradeoff between having processes sit idle for some bricks versus having a long wallclock time for others.  If we changed the organization of the data, then it would be easier to load balance.  Note that I *AM* already load balancing the bricks assigned to each process *group* based on the number of spectra in each brick.  However, each process group still has bricks of different sizes.  Rather than just going down this road, we should find a "good" format for the data from a computational perspective, and then re-pack it later for usability.

**Comment by sbailey** [comment]
There is some mismatch on the redshift performance compared to the oak run.  Oak bricks varied from 1 to 513 targets, but still finished within 48 minutes (including 7 minutes of python startup time) on 154 nodes.  This is way better than 343 minutes for 600 targets on the largest brick for pine.  Pine is about 3x more exposures than oak, but only used 168 nodes compared to 154 for oak.  Is it possible that the pipeline planning grouped multiple large bricks together?

**Comment by tskisner** [comment]
I agree that something has changed.  I am currently running with a process group of size 24- I think Oak was using 48.  Has anything internally changed in redmonster as well?  grepping through the logs with:

   `$>  grep 'Time to find' zbest-*.log | grep QSO`

I see that the initial fit of the QSO template takes about 260s per fiber.  Has it always taken that long?  For some large bricks, some processes have 13-14 fibers:

   `zbest-3211p055.log:DEBUG:redmonster.py:108:__init__: Time to find the redshifts of 14 fibers for template ndArch-QSO-V003.fits =3780.255076 sec `

So we could increase the number of processes working on a single brick, but if that group of processes is assigned a "small" brick then many of those processes would sit idle.  That seems silly.  Also, shuffling 100GB of data between formats should not require an hour or storing the whole thing in RAM.

**Comment by sbailey** [comment]
oak had different logging, but for comparison:
```
INFO fitted fiber 7/10, chi2_null=8157.648390, 125 templates in /project/projectdirs//desi/spectro/templates
/redmonster/v0.1/ndArch-QSO-V003.fits, npoly=2, using 1 procs in 18.204429 sec
```
i.e. one fiber took 18 sec for ndArch-QSO-V003.fits, while pine took 3780/14 = 270 sec (!).  We did have the python 3 upgrade in there, but I'm not aware of any change that should make that dramatic a difference in timing.

oak logs in /project/projectdirs/desi/spectro/redux/oak1/run/logs

**Comment by sbailey** [comment]
Perhaps related to slow redmonster performance: many of the /project/projectdirs/desi/spectro/redux/pine_214ae69/bricks/ files are corrupted due to #312 (corrupted brick file generation).  That issue was fixed in PR #314.  If redmonster is working off of corrupted brick files it could have unpredictable behavior.

**Comment by tskisner** [comment]
Interesting.  Ok, my retry of the redmonster run is still queued, I'll put it on hold and rebuild the bricks.
